File: 188-best-time-to-buy-and-sell-stock-iv/188-best-time-to-buy-and-sell-stock-iv.java
class Solution {
    public int maxProfit(int k, int[] prices) {
          return fun(prices,0, 1,k, new HashMap<String, Integer>());
    }
    
     public int fun(int [] prices, int cd, int canBuy, int transaction, HashMap<String, Integer> memo){
        
        
        if(cd>=prices.length || transaction==0)return 0;
        
        
        
      //  String currkey=Integer.toString(cd)+'_'+Integer.toString(canBuy)+'_'+Integer.toString(transaction);
         String currkey= ""+cd+"_"+canBuy+"_"+transaction;
             
             
         if(memo.containsKey(currkey))return memo.get(currkey);
        
        int idle=fun(prices,cd+1,canBuy,transaction,memo);
        
       
        if(canBuy==1)
        {
            int buy=-prices[cd]+fun(prices,cd+1,0,transaction,memo);
             memo.put(currkey, Math.max(idle,buy));
            return Math.max(idle,buy);

        }   
        
      
        else{
           int sell=prices[cd]+fun(prices,cd+1,1,transaction-1,memo);
             memo.put(currkey, Math.max(idle,sell));
            return Math.max(idle,sell);
        }
        
        // return memo.get(currkey);
       
    }
}

File: 188-best-time-to-buy-and-sell-stock-iv/188-best-time-to-buy-and-sell-stock-iv.cpp
class Solution {
public:
    int maxProfit(int k, vector<int>& prices) {
      vector<vector<vector<int>>>dp(prices.size()+1,vector<vector<int>>(2,vector<int>(k+1,-1)));
        
        //prices.size() canBuy kTransaction
        int currDay=0;
        int canBuy=1;
        // int kTrans=1;
        return maxEarn(dp,prices,currDay, canBuy, k,prices.size());
        
    }
    
    int maxEarn(vector<vector<vector<int>>>&dp,vector<int>& prices, int currDay, int canBuy, int kTrans,int n ){
        
       if(currDay>=n)return 0;
        
        if(kTrans<=0)return 0;
        
        if(dp[currDay][canBuy][kTrans]!=-1)
            return dp[currDay][canBuy][kTrans];
        
        if(canBuy==1){
            int buy=-prices[currDay]+maxEarn(dp,prices,currDay+1,0,kTrans,n);
            int idle=maxEarn(dp,prices,currDay+1,canBuy,kTrans,n);
            
            dp[currDay][canBuy][kTrans]=max(buy,idle);
        }else if(canBuy==0){
            
             int sell=prices[currDay]+maxEarn(dp,prices,currDay+1,1,kTrans-1,n);
            int idle=maxEarn(dp,prices,currDay+1,canBuy,kTrans,n);
              dp[currDay][canBuy][kTrans]=max(sell,idle);
        }
       return  dp[currDay][canBuy][kTrans];
    }
    
};